logtail/logtail_test.go

```diff
@@ -23,29 +28,245 @@ func TestFastShutdown(t *testing.T) {
+			ts.uploaded <- 0
 		}))
-	defer testServ.Close()
 
-	l := NewLogger(Config{BaseURL: testServ.URL}, t.Logf)
-	for i := 1; i < 10; i++ {
+	return &ts
+}
+
+func (ts *LogtailTestServer) HandleStartLoggingMessage() {
+	// There is always an initial "logtail started" message
+	<-ts.uploaded
+	ts.mu.Lock()
+	if !strings.Contains(ts.bodytext, "started") {
+		panic("unknown start logging statement:" + ts.bodytext)
+	}
+	ts.mu.Unlock()
+}
+
+func TestDrainPendingMessages(t *testing.T) {
+	logLines := 3
+	ts := NewLogtailTestServer(logLines)
+	l := NewLogger(Config{BaseURL: ts.srv.URL}, t.Logf)
+	ts.HandleStartLoggingMessage()
+
+	for i := 0; i < logLines; i++ {
 		l.Write([]byte("log line"))
 	}
 
-	l.Shutdown(ctx)
-	cancel()
-	if uploads == 0 {
-		t.Error("no log uploads")
+	// all of the "log line" messages usually arrive at once, but poll if needed.
+	for i := 0; i <= logLines; i++ {
+		<-ts.uploaded
+		ts.mu.Lock()
+		cnt := strings.Count(ts.bodytext, "log line")
+		ts.mu.Unlock()
+		if cnt == logLines {
+			break
+		}
 	}
+
+	err := l.Shutdown(ts.ctx)
+	if err != nil {
+		t.Error(err)
+	}
+	ts.srv.Close()
+	ts.cancel()
+
+	ts.mu.Lock()
+	if strings.Count(ts.bodytext, "log line") != logLines {
+		t.Errorf("got %q; want: %d copies of %q", ts.bodytext, logLines, "log line")
+	}
+	ts.mu.Unlock()
+}
+
+func TestEncodeAndUploadMessages(t *testing.T) {
+	ts := NewLogtailTestServer(1)
+	l := NewLogger(Config{BaseURL: ts.srv.URL}, t.Logf)
+	ts.HandleStartLoggingMessage()
+
+	tests := []struct {
+		name string
+		log  string
+		want string
+	}{
+		{
+			"plain text",
+			"log line",
+			"log line",
+		},
+		{
+			"simple JSON",
+			`{"text": "log line"}`,
+			"log line",
+		},
+	}
+
+	for _, tt := range tests {
+		io.WriteString(l, tt.log)
+		<-ts.uploaded
+
+		data := make(map[string]interface{})
+		ts.mu.Lock()
+		err := json.Unmarshal(ts.body, &data)
+		ts.mu.Unlock()
+		if err != nil {
+			t.Error(err)
+		}
+
+		got, ok := data["text"]
+		if ok {
+			if got != tt.want {
+				t.Errorf("%s: got %q; want %q", tt.name, got.(string), tt.want)
+			}
+		} else {
+			t.Errorf("no text in: %v", data)
+		}
+
+		ltail, ok := data["logtail"]
+		if ok {
+			logtailmap := ltail.(map[string]interface{})
+			_, ok = logtailmap["client_time"]
+			if !ok {
+				t.Errorf("%s: no client_time present", tt.name)
+			}
+		} else {
+			t.Errorf("%s: no logtail map present", tt.name)
+		}
+	}
+
+	err := l.Shutdown(ts.ctx)
+	if err != nil {
+		t.Error(err)
+	}
+	ts.srv.Close()
+	ts.cancel()
+}
+
+func TestEncodeSpecialCases(t *testing.T) {
+	ts := NewLogtailTestServer(1)
+	l := NewLogger(Config{BaseURL: ts.srv.URL}, t.Logf)
+	ts.HandleStartLoggingMessage()
+
+	// -------------------------------------------------------------------------
+
+	// JSON log message already contains a logtail field.
+	io.WriteString(l, `{"logtail": "LOGTAIL", "text": "text"}`)
+	<-ts.uploaded
+	data := make(map[string]interface{})
+	ts.mu.Lock()
+	err := json.Unmarshal(ts.body, &data)
```

Comment:
Rather than manually Lock/Unlock ts.mu when you need ts.body, I suggest adding a ts method to retrieve body, which does the locking for you.

Reply:
In fact, it looks like you typically block on uploaded and then use a mutex to read body/bodytext. What if instead you passed a struct containing body and bodytext over the uploaded channel? Perhaps you could get rid of the mutexes entirely.

Reply:
Done. passed just the bytes[] over the channel, there is only one test which needs to concatenate multiple log messages as strings so made that test handle its own concatenation.